File: 00543. Diameter of Binary Tree/00543.c
int dfs(struct TreeNode *root, int *max) {
    int l = 0, r = 0;
    
    if (root->left != NULL) {
        l = dfs(root->left, max);
    }
    
    if (root->right != NULL) {
        r = dfs(root->right, max);
    }
    
    if (*max < l + r) {
        *max = l + r;
    }
    
    return (l > r) ? l + 1 : r + 1;
}

int diameterOfBinaryTree(struct TreeNode *root) {
    int max = 0;
    
    dfs(root, &max);
    
    return max;
}

File: 00543. Diameter of Binary Tree/00543.go
// https://leetcode.com/problems/diameter-of-binary-tree

func helper(root *TreeNode, ans *int) int {
    if root == nil {
        return 0
    }

    l := helper(root.Left, ans)
    r := helper(root.Right, ans)
    if *ans < l+r {
        *ans = l+r
    }

    return max(l, r) + 1
}

func diameterOfBinaryTree(root *TreeNode) int {
    ans := 0
    helper(root, &ans)
    return ans
}
